common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/StreamStatesCleanupSuite.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.shuffle;
+
+import java.nio.ByteBuffer;
+
+import io.netty.channel.Channel;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import org.apache.spark.network.buffer.ManagedBuffer;
+import org.apache.spark.network.buffer.NioManagedBuffer;
+import org.apache.spark.network.client.RpcResponseCallback;
+import org.apache.spark.network.client.TransportClient;
+import org.apache.spark.network.client.TransportResponseHandler;
+import org.apache.spark.network.server.OneForOneStreamManager;
+import org.apache.spark.network.server.RpcHandler;
+import org.apache.spark.network.shuffle.protocol.OpenBlocks;
+
+public class StreamStatesCleanupSuite {
+
+  @Test
+  public void testStreamsAreRemovedCorrectly() {
+    OneForOneStreamManager streamManager = new OneForOneStreamManager();
+    ExternalShuffleBlockResolver blockResolver = mock(ExternalShuffleBlockResolver.class);
+    TransportClient reverseClient
+      = new TransportClient(mock(Channel.class), mock(TransportResponseHandler.class));
+    RpcHandler handler = new ExternalShuffleBlockHandler(streamManager, blockResolver);
+
+    ManagedBuffer block0Marker = new NioManagedBuffer(ByteBuffer.wrap(new byte[3]));
+    ManagedBuffer block1Marker = new NioManagedBuffer(ByteBuffer.wrap(new byte[7]));
+    when(blockResolver.getBlockData("app0", "exec1", 0, 0, 0))
+      .thenReturn(block0Marker);
+    when(blockResolver.getBlockData("app0", "exec1", 0, 0, 1))
+      .thenReturn(block1Marker);
+    ByteBuffer openBlocks = new OpenBlocks("app0", "exec1",
+      new String[]{"shuffle_0_0_0", "shuffle_0_0_1"})
+      .toByteBuffer();
+
+    RpcResponseCallback callback = mock(RpcResponseCallback.class);
+
+    // Open blocks
+    handler.receive(reverseClient, openBlocks, callback);
```

Comment:
Assert stream count after receiving an `OpenBlocks` message.

Reply:
@Ngone51 Thanks a lot for comment and review!

> A minor concern is:
> Since we always register the channel when we receive the `FetchChunkRequest` in the past, so the associated channel with the stream will aways be correct. And for this change, if user call `sendRpc(OpenBlocks)` and `fetchChunk` with different client, then, we may record the wrong channel(which is the `OpenBlocks` chanel, rather than `fetchChunk` channel) with the stream which is being fetched, which may fall into the same issue as this pr fixed. So, I think we should leave some comment above `fetchChunk` to notify user that this method _must_ use the same client with the `OpenBlocks`, otherwise, there may be potential memory leak.

I had the same concern too. I added some comment at `TransportClient.fetchChunk` .